README.md
RIO
======
[![Build Status](https://travis-ci.org/albertosantini/node-rio.png)](https://travis-ci.org/albertosantini/node-rio)
[![NPM version](https://badge.fury.io/js/rio.png)](http://badge.fury.io/js/rio)
[![NGN Dependencies](https://david-dm.org/albertosantini/node-rio.png)](https://david-dm.org/albertosantini/node-rio)

RIO, R Input Output, connects an app to [Rserve](http://www.rforge.net/Rserve/),
a TCP/IP server which allows other programs to use facilities of [R](http://www.r-project.org).

It supports double, double array, integer, integer array, string, string array,
boolean, boolean array objects and raw vector (images or files).

It supports also the plain text and crypted authentication, if Rserve is
configured for that capability.

The main goal is to pass a string containing a script call using a JSON object
as parameter. Then, inside the script, using `RJSONIO` or `jsonlite` package,
deserializing the JSON object, calling a method, serializing the response and
returning to Node.js.

Examples
========

    var rio = require("rio");

    rio.e({command: "pi / 2 * 2"});
    rio.e({command: "c(1, 2)"});
    rio.e({command: "as.character('Hello World')"});
    rio.e({command: "c('a', 'b')"});
    rio.e({command: "Sys.sleep(5); 11"})

    rio.$e({
        command: "pi / 2 * 2"
    }).then(function (res) {
        console.log(res);
    });

    rio.e({
        command: "2 + 2"
    }).e({
        command: "3 + 3"
    });

See `examples` directory.

- `ex1`: Getting started with `evaluate` api.
- `ex2`: How to evaluate a `filename` and `entrypoint`.
- `ex3`: How to evaluate a `filename` and `host`.
- `ex4`: An example with utf-8 chars.
- `ex5`: How to retrieve a plot.
- `ex6`: How to call functions already loaded in R session.
- `ex7`: An example with large data packet.
- `ex8`: An example with `evaluateDefer` api.
- `ex9`: An example chaining `evaluate` api.
- `ex10`: How to evaluate a matrix, using JSON serialization.

Installation
============

[![NPM](https://nodei.co/npm/rio.png?downloads=true&downloadRank=true)](https://nodei.co/npm/rio/)
[![NPM](https://nodei.co/npm-dl/rio.png)](https://nodei.co/npm/rio/)

To install with [npm](http://github.com/isaacs/npm):

    npm install rio

Tested with Node.js 5.x and Rserve 1.7.3, on Windows 10 64bit with R 3.2.4
and on Debian Jessie (USB armory) with R 3.1.1.

Don't forget to start [Rserve](http://cran.r-project.org/web/packages/Rserve/).
For instance, from R console, after installing the package Rserve:

    require("Rserve")
    Rserve()

To shutdown the server from R console:

    require("RSclient")
    c <- RSconnect()
    RSshutdown(c)

Methods
=======

evaluate(config) - e(config)
----------------------------

Evaluate a command, connecting to Rserve, executing the command and then
disconnecting. The result is passed to the callback.

The defaults for the options parameter:

    config = {
        command: "",
        filename: "",

        entrypoint: "",
        data: {},

        callback: function (err, res) {
            if (!err) {
                console.log(res);
            } else {
                console.log("Rserve call failed. " + err);
            }
        },

        host = "127.0.0.1",
        port = "6311",
        path = undefined,

        user = "anon",
        password = "anon"
    }

- `command` OR `filename` OR `entrypoint` need to be filled.
Otherwise it is missing the evaluation object.

- if `command` AND  `filename` AND `entrypoint` are empty then error.
As above, said in different way.

- `command` AND `filename` are exclusive: if both are not empty then error.
Otherwise what does rio evaluate, command or filename?

- if `command` AND  `filename` are empty then `entrypoint` is mandatory.
This is the case when rio evaluates a function defined on R side.

- `host` AND `path` are exclusive.
rio needs to choose beetween net socket or unix socket transport.

When `filename` is filled, rio loads the content of a R file, calling
finally an `entrypoint`, passing `data`.

    config = {
        filename: "foo.R",
        entrypoint: "main", // entrypoint is called
        data: { foo: "bar" } // data is stringified and passed to entrypoint
    }

When `entrypoint` is filled, finally passing `data`, it is used when we
need to call a function defined in Rserve instance.

    config = {
        entrypoint: "echo",
        data: ["test", "data"],
        callback: printEcho
    }

evaluateDefer(config) - $e(config)
----------------------------------

Evaluate a command, returning a promise: config options is the same as
`evaluate`.

shutdown(options)
-----------------

Sends the `CMD_shutdown` command to the Rserve server. Options are the same as
for `evaluate`.

enableDebug(isDebug)
-----------

It enables debugging mode, printing the packet and logging messages on client
side.

You may start also a Rserve instance in debugging mode with following commands
(on Windows box with Git Bash Shell):

```
export R_PATH=/c/My/Programs/R
export PATH=$PATH:$R_PATH/bin/x64

$R_PATH/library/Rserve/libs/x64/Rserve_d.exe --
```

Set your paths accordingly.

enableRecordMode(isRecordMode, options)
----------------

It enables record mode, dumping the incoming data to a file specified in the
options.

    options = {
        fileName: "node-rio-dump.bin"
    }

It is useful to record a Rserve session to replay it in an environment without
[Rserve](http://cran.r-project.org/web/packages/Rserve/) (for example
[Travis CI](https://travis-ci.org/)). For instance,

```
> var rio=require("./index.js")
undefined
> rio.enableRecordMode(true, {fileName: "test/dump/integer-test.bin"});
undefined
> rio.evaluate({command: "as.integer(3)"})
undefined
> 3
(^C again to quit)
```

Then, you need to export the variable `CI` to emulate CI environment:
`export CI=true`

Eventually `npm test`.

enablePlaybackMode(isPlaybackMode, options)
------------------

It enables playback mode, reading a dump file instead connecting to the server.

    options = {
        fileName: "node-rio-dump.bin"
    }


Contributors
============

```
project  : node-rio
repo age : 4 years, 8 months
active   : 122 days
commits  : 272
files    : 61
authors  :
  243  icebox                  89.3%
    8  Alberto Santini         2.9%
    7  Manuel Santillan        2.6%
    6  albertosantini          2.2%
    3  Karthik Madathil        1.1%
    2  Anand Patil             0.7%
    1  Alex Proca              0.4%
    1  Farrin Reid             0.4%
    1  Koichiro Sobue          0.4%
```


CHANGELOG.md
2.4.1 / 2016-03-31
==================

* Added matrix example using JSON serialization.
* Added check params test.
* Updated deps.

2.4.0 / 2015-12-30
==================

* Added chaining support.
* Changed from vows tests to tape ones.

2.3.0 / 2015-12-27
==================

* Added evaluate defer API: $e.

2.2.0 / 2015-12-07
==================

* Added support for crypted authentication.

2.1.0 / 2015-11-23
==================

* Added support for large packet.

2.0.1 / 2015-11-22
==================

*  Fixed getting started examples in README.

2.0.0 / 2015-11-22
==================

* Interface changes and implementation refactoring, no new features.

1.4.0 / 2014-11-26
==================

* Added option to call functions already loaded in R session.
* Used eslint (vs. jshint).

1.3.2 / 2014-11-14
==================

* Added support for integer and integer array.
* Added info to record a Rserve session for the tests.
* Updated hexy 0.2.7, vows 0.8.0.

1.3.1 / 2014-09-08
==================

* Fixed #16 - End-of-message padding seems wrong at least on Mac OS X.

1.3.0 / 2014-04-27
==================

* Fixed #13 - Error strings/network errors propagation.

1.2.1 / 2014-04-10
==================

* Fix #12 - SIGPIPE signal on remote Rserve.

1.2.0 / 2013-11-12
==================

* Added bufferAndEval.
* Added login tests.

1.1.0 / 2013-11-06
==================

* Added boolean support.
* Added eval response and data type check.
* Added eval errors tests.

1.0.0 / 2013-10-27
==================

* After a few cosmetic updates, `rio` is ready for 1.0.0.

0.10.0 / 2013-08-18
===================

* Added transfer file support.
* Fixed callback err value if sexp type is not implemented.
* Added contributing doc.
* Added npm stats, dependencies and version badges.

0.9.0 / 2013-06-18
==================

* Added Rserve connection on socket as well as tcp/ip port.
* Added shutdown command.

0.8.1 / 2013-05-23
==================

* Fixed utf8 support in the response.
* Added keepalive for long running processes.

0.8.0 / 2013-04-23
==================

* Added utf8 support.

0.7.0 / 2013-03-13
==================

* Added tests without needing Rserve.
* Added the err argument in the callback.

0.6.2 / 2013-03-08
==================

* Removed jspack dependency.

0.6.1 / 2012-06-25
==================

* Fixed History doc for unintentional deletion.
* Changed sys to util require for compatibility with 0.6.x and higher.

0.6.0 / 2011-11-20
==================

* Refactored the data handler using node-binary.

0.5.6 / 2011-11-17
==================

* Fixed the copy in splitted packet.

0.5.5 / 2011-11-17
==================

* Fixed mangling a byte in the splitted packet.
* 0.5.4 tag was 0.5.3 one.

0.5.3 / 2011-11-17
==================

* Fixed data packet splitted.

0.5.2 / 2011-11-15
==================

* Fixed data packet with length 16 bytes.

0.5.1 / 2011-09-26
==================

* Fixed a bug in sending commands loop.

0.5.0 / 2011-09-25
==================

* Fixed bugs for testing with unix Rserve instance.

0.4.0 / 2011-09-15
==================

* Added enableDebug sourceAndEval methods.
* Fixed bugs.

0.3.1 / 2011-08-11
==================

* Fixed a typo in the README.

0.3.0 / 2011-08-10
==================

* Implemented plaintext authentication.

0.2.0 / 2011-08-08
==================

* Fixed an issue with the length of the packet because string and buffer were
mixed.

0.1.0 / 2011-08-04
==================

* Initial release.


lib/crypt.js
/*eslint-env es6 */
/*eslint-disable  */

module.exports = descrypt;

// https://gist.githubusercontent.com/Tyilo/5954072/raw/ae62af0d577569b6de71b9b642aeb0a5bd6861d4/DES%2520crypt.js

/* JavaScript password hash generator.
 * $Id: pwd.js,v 1.5 2004/10/09 09:41:38 emikulic Exp $
 *
 * Copyright (c) 2004, Emil Mikulic.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the author nor the names of other contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS''
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
 * PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

const ascii64 =
"./0123456789ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz";

/* Emil's adaptation of crypt() from FreeSec: libcrypt
 * Taken from $OpenBSD: crypt.c,v 1.18 2003/08/12 01:22:17 deraadt Exp $
 * Original license:
 * Copyright (c) 1994 David Burren
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 4. Neither the name of the author nor the names of other contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE AUTHOR AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE AUTHOR OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * The JavaScript adaptation is copyright (c) 2004 Emil Mikulic.
 */

function ascii_to_bin(ch)
{
 var lz = "z".charCodeAt(0),
     la = "a".charCodeAt(0),
     uz = "Z".charCodeAt(0),
     ua = "A".charCodeAt(0),
     ni = "9".charCodeAt(0),
     dt = ".".charCodeAt(0);

 if (ch >  lz) return 0;
 if (ch >= la) return (ch - la + 38);
 if (ch >  uz) return 0;
 if (ch >= ua) return (ch - ua + 12);
 if (ch >  ni) return 0;
 if (ch >= dt) return (ch - dt);
 return 0;
}

const des_IP = [
        58, 50, 42, 34, 26, 18, 10,  2, 60, 52, 44, 36, 28, 20, 12,  4,
        62, 54, 46, 38, 30, 22, 14,  6, 64, 56, 48, 40, 32, 24, 16,  8,
        57, 49, 41, 33, 25, 17,  9,  1, 59, 51, 43, 35, 27, 19, 11,  3,
        61, 53, 45, 37, 29, 21, 13,  5, 63, 55, 47, 39, 31, 23, 15,  7
];

const des_key_perm = [
        57, 49, 41, 33, 25, 17,  9,  1, 58, 50, 42, 34, 26, 18,
        10,  2, 59, 51, 43, 35, 27, 19, 11,  3, 60, 52, 44, 36,
        63, 55, 47, 39, 31, 23, 15,  7, 62, 54, 46, 38, 30, 22,
        14,  6, 61, 53, 45, 37, 29, 21, 13,  5, 28, 20, 12,  4
];

const des_key_shifts = [ 1, 1, 2, 2, 2, 2, 2, 2, 1, 2, 2, 2, 2, 2, 2, 1 ];

const des_comp_perm = [
        14, 17, 11, 24,  1,  5,  3, 28, 15,  6, 21, 10,
        23, 19, 12,  4, 26,  8, 16,  7, 27, 20, 13,  2,
        41, 52, 31, 37, 47, 55, 30, 40, 51, 45, 33, 48,
        44, 49, 39, 56, 34, 53, 46, 42, 50, 36, 29, 32
];

const des_sbox = [
[
        14,  4, 13,  1,  2, 15, 11,  8,  3, 10,  6, 12,  5,  9,  0,  7,
         0, 15,  7,  4, 14,  2, 13,  1, 10,  6, 12, 11,  9,  5,  3,  8,
         4,  1, 14,  8, 13,  6,  2, 11, 15, 12,  9,  7,  3, 10,  5,  0,
        15, 12,  8,  2,  4,  9,  1,  7,  5, 11,  3, 14, 10,  0,  6, 13
], [
        15,  1,  8, 14,  6, 11,  3,  4,  9,  7,  2, 13, 12,  0,  5, 10,
         3, 13,  4,  7, 15,  2,  8, 14, 12,  0,  1, 10,  6,  9, 11,  5,
         0, 14,  7, 11, 10,  4, 13,  1,  5,  8, 12,  6,  9,  3,  2, 15,
        13,  8, 10,  1,  3, 15,  4,  2, 11,  6,  7, 12,  0,  5, 14,  9
], [
        10,  0,  9, 14,  6,  3, 15,  5,  1, 13, 12,  7, 11,  4,  2,  8,
        13,  7,  0,  9,  3,  4,  6, 10,  2,  8,  5, 14, 12, 11, 15,  1,
        13,  6,  4,  9,  8, 15,  3,  0, 11,  1,  2, 12,  5, 10, 14,  7,
         1, 10, 13,  0,  6,  9,  8,  7,  4, 15, 14,  3, 11,  5,  2, 12
], [
         7, 13, 14,  3,  0,  6,  9, 10,  1,  2,  8,  5, 11, 12,  4, 15,
        13,  8, 11,  5,  6, 15,  0,  3,  4,  7,  2, 12,  1, 10, 14,  9,
        10,  6,  9,  0, 12, 11,  7, 13, 15,  1,  3, 14,  5,  2,  8,  4,
         3, 15,  0,  6, 10,  1, 13,  8,  9,  4,  5, 11, 12,  7,  2, 14
], [
         2, 12,  4,  1,  7, 10, 11,  6,  8,  5,  3, 15, 13,  0, 14,  9,
        14, 11,  2, 12,  4,  7, 13,  1,  5,  0, 15, 10,  3,  9,  8,  6,
         4,  2,  1, 11, 10, 13,  7,  8, 15,  9, 12,  5,  6,  3,  0, 14,
        11,  8, 12,  7,  1, 14,  2, 13,  6, 15,  0,  9, 10,  4,  5,  3
], [
        12,  1, 10, 15,  9,  2,  6,  8,  0, 13,  3,  4, 14,  7,  5, 11,
        10, 15,  4,  2,  7, 12,  9,  5,  6,  1, 13, 14,  0, 11,  3,  8,
         9, 14, 15,  5,  2,  8, 12,  3,  7,  0,  4, 10,  1, 13, 11,  6,
         4,  3,  2, 12,  9,  5, 15, 10, 11, 14,  1,  7,  6,  0,  8, 13
], [
         4, 11,  2, 14, 15,  0,  8, 13,  3, 12,  9,  7,  5, 10,  6,  1,
        13,  0, 11,  7,  4,  9,  1, 10, 14,  3,  5, 12,  2, 15,  8,  6,
         1,  4, 11, 13, 12,  3,  7, 14, 10, 15,  6,  8,  0,  5,  9,  2,
         6, 11, 13,  8,  1,  4, 10,  7,  9,  5,  0, 15, 14,  2,  3, 12
], [
        13,  2,  8,  4,  6, 15, 11,  1, 10,  9,  3, 14,  5,  0, 12,  7,
         1, 15, 13,  8, 10,  3,  7,  4, 12,  5,  6, 11,  0, 14,  9,  2,
         7, 11,  4,  1,  9, 12, 14,  2,  0,  6, 10, 13, 15,  3,  5,  8,
         2,  1, 14,  7,  4, 10,  8, 13, 15, 12,  9,  0,  3,  5,  6, 11
]];

const des_pbox = [
        16,  7, 20, 21, 29, 12, 28, 17,  1, 15, 23, 26,  5, 18, 31, 10,
         2,  8, 24, 14, 32, 27,  3,  9, 19, 13, 30,  6, 22, 11,  4, 25
];

const des_bits32 = [
        0x80000000, 0x40000000, 0x20000000, 0x10000000,
        0x08000000, 0x04000000, 0x02000000, 0x01000000,
        0x00800000, 0x00400000, 0x00200000, 0x00100000,
        0x00080000, 0x00040000, 0x00020000, 0x00010000,
        0x00008000, 0x00004000, 0x00002000, 0x00001000,
        0x00000800, 0x00000400, 0x00000200, 0x00000100,
        0x00000080, 0x00000040, 0x00000020, 0x00000010,
        0x00000008, 0x00000004, 0x00000002, 0x00000001
];

const bits28 = [
        0x08000000, 0x04000000, 0x02000000, 0x01000000,
        0x00800000, 0x00400000, 0x00200000, 0x00100000,
        0x00080000, 0x00040000, 0x00020000, 0x00010000,
        0x00008000, 0x00004000, 0x00002000, 0x00001000,
        0x00000800, 0x00000400, 0x00000200, 0x00000100,
        0x00000080, 0x00000040, 0x00000020, 0x00000010,
        0x00000008, 0x00000004, 0x00000002, 0x00000001
];

const bits24 = [
        0x00800000, 0x00400000, 0x00200000, 0x00100000,
        0x00080000, 0x00040000, 0x00020000, 0x00010000,
        0x00008000, 0x00004000, 0x00002000, 0x00001000,
        0x00000800, 0x00000400, 0x00000200, 0x00000100,
        0x00000080, 0x00000040, 0x00000020, 0x00000010,
        0x00000008, 0x00000004, 0x00000002, 0x00000001
];

const des_bits8 = [ 0x80, 0x40, 0x20, 0x10, 0x08, 0x04, 0x02, 0x01 ];

var     u_sbox = new Array(8),
        m_sbox = new Array(4),
        init_perm = new Array(64),
        final_perm = new Array(64),
        inv_key_perm = new Array(64),
        u_key_perm = new Array(56),
        inv_comp_perm = new Array(56),
        ip_maskl = new Array(8),
        ip_maskr = new Array(8),
        fp_maskl = new Array(8),
        fp_maskr = new Array(8),
        un_pbox = new Array(32),
        psbox = new Array(4),
        key_perm_maskl = new Array(8),
        key_perm_maskr = new Array(8),
        comp_maskl = new Array(8),
        comp_maskr = new Array(8);

function des_init()
{
        /*
         * Invert the S-boxes, reordering the input bits.
         */
        for (i = 0; i < 8; i++) {
                u_sbox[i] = new Array(64);
                for (j = 0; j < 64; j++) {
                        b = (j & 0x20) | ((j & 1) << 4) | ((j >> 1) & 0xf);
                        u_sbox[i][j] = des_sbox[i][b];
                }
        }

        /*
         * Convert the inverted S-boxes into 4 arrays of 8 bits.
         * Each will handle 12 bits of the S-box input.
         */
        for (b = 0; b < 4; b++) {
                m_sbox[b] = new Array(4096);
                for (i = 0; i < 64; i++)
                        for (j = 0; j < 64; j++)
                                m_sbox[b][(i << 6) | j] =
                                        (u_sbox[(b << 1)][i] << 4) |
                                        u_sbox[(b << 1) + 1][j];
        }

        /*
         * Set up the initial & final permutations into a useful form, and
         * initialise the inverted key permutation.
         */
        for (i = 0; i < 64; i++) {
                init_perm[final_perm[i] = des_IP[i] - 1] = i;
                inv_key_perm[i] = 255;
        }

        /*
         * Invert the key permutation and initialise the inverted key
         * compression permutation.
         */
        for (i = 0; i < 56; i++) {
                u_key_perm[i] = des_key_perm[i] - 1;
                inv_key_perm[des_key_perm[i] - 1] = i;
                inv_comp_perm[i] = 255;
        }

        /*
         * Invert the key compression permutation.
         */
        for (i = 0; i < 48; i++) {
                inv_comp_perm[des_comp_perm[i] - 1] = i;
        }

        /*
         * Set up the OR-mask arrays for the initial and final permutations,
         * and for the key initial and compression permutations.
         */
        for (k = 0; k < 8; k++) {
                ip_maskl[k] = new Array(256);
                ip_maskr[k] = new Array(256);
                fp_maskl[k] = new Array(256);
                fp_maskr[k] = new Array(256);
                key_perm_maskl[k] = new Array(128);
                key_perm_maskr[k] = new Array(128);
                comp_maskl[k] = new Array(128);
                comp_maskr[k] = new Array(128);

                for (i = 0; i < 256; i++) {
                        ip_maskl[k][i] = 0;
                        ip_maskr[k][i] = 0;
                        fp_maskl[k][i] = 0;
                        fp_maskr[k][i] = 0;
                        for (j = 0; j < 8; j++) {
                                inbit = 8 * k + j;
                                if (i & des_bits8[j]) {
                                        if ((obit = init_perm[inbit]) < 32)
                                                ip_maskl[k][i] |=
                                                  des_bits32[obit];
                                        else
                                                ip_maskr[k][i] |=
                                                  des_bits32[obit-32];

                                        if ((obit = final_perm[inbit]) < 32)
                                                fp_maskl[k][i] |=
                                                  des_bits32[obit];
                                        else
                                                fp_maskr[k][i] |=
                                                  des_bits32[obit - 32];
                                }
                        }
                }
                for (i = 0; i < 128; i++) {
                        key_perm_maskl[k][i] = 0;
                        key_perm_maskr[k][i] = 0;
                        for (j = 0; j < 7; j++) {
                                inbit = 8 * k + j;
                                if (i & des_bits8[j + 1]) {
                                        if ((obit = inv_key_perm[inbit]) == 255)
                                                continue;
                                        if (obit < 28)
                                                key_perm_maskl[k][i] |=
                                                  bits28[obit];
                                        else
                                                key_perm_maskr[k][i] |=
                                                  bits28[obit - 28];
                                }
                        }
                        comp_maskl[k][i] = 0;
                        comp_maskr[k][i] = 0;
                        for (j = 0; j < 7; j++) {
                                inbit = 7 * k + j;
                                if (i & des_bits8[j + 1]) {
                                        if ((obit=inv_comp_perm[inbit]) == 255)
                                                continue;
                                        if (obit < 24)
                                                comp_maskl[k][i] |=
                                                  bits24[obit];
                                        else
                                                comp_maskr[k][i] |=
                                                  bits24[obit - 24];
                                }
                        }
                }
        }

        /*
         * Invert the P-box permutation, and convert into OR-masks for
         * handling the output of the S-box arrays setup above.
         */
        for (i = 0; i < 32; i++)
                un_pbox[des_pbox[i] - 1] = i;

        for (b = 0; b < 4; b++) {
                psbox[b] = new Array(256);
                for (i = 0; i < 256; i++) {
                        psbox[b][i] = 0;
                        for (j = 0; j < 8; j++) {
                                if (i & des_bits8[j])
                                        psbox[b][i] |=
                                          des_bits32[un_pbox[8 * b + j]];
                        }
                }
        }
}

var en_keysl = new Array(16),
    en_keysr = new Array(16);

function des_setkey(key)
{
 var rawkey0, rawkey1, k0, k1;

 rawkey0 = (key[0] << 24) |
           (key[1] << 16) |
           (key[2] <<  8) |
           (key[3] <<  0);

 rawkey1 = (key[4] << 24) |
           (key[5] << 16) |
           (key[6] <<  8) |
           (key[7] <<  0);

 /* Do key permutation and split into two 28-bit subkeys. */
 k0 = key_perm_maskl[0][rawkey0 >>> 25]
    | key_perm_maskl[1][(rawkey0 >>> 17) & 0x7f]
    | key_perm_maskl[2][(rawkey0 >>> 9) & 0x7f]
    | key_perm_maskl[3][(rawkey0 >>> 1) & 0x7f]
    | key_perm_maskl[4][rawkey1 >>> 25]
    | key_perm_maskl[5][(rawkey1 >>> 17) & 0x7f]
    | key_perm_maskl[6][(rawkey1 >>> 9) & 0x7f]
    | key_perm_maskl[7][(rawkey1 >>> 1) & 0x7f];
 k1 = key_perm_maskr[0][rawkey0 >>> 25]
    | key_perm_maskr[1][(rawkey0 >>> 17) & 0x7f]
    | key_perm_maskr[2][(rawkey0 >>> 9) & 0x7f]
    | key_perm_maskr[3][(rawkey0 >>> 1) & 0x7f]
    | key_perm_maskr[4][rawkey1 >>> 25]
    | key_perm_maskr[5][(rawkey1 >>> 17) & 0x7f]
    | key_perm_maskr[6][(rawkey1 >>> 9) & 0x7f]
    | key_perm_maskr[7][(rawkey1 >>> 1) & 0x7f];

 /* Rotate subkeys and do compression permutation. */
 var shifts = 0, round;
 for (round = 0; round < 16; round++) {
         var t0, t1;

         shifts += des_key_shifts[round];

         t0 = (k0 << shifts) | (k0 >>> (28 - shifts));
         t1 = (k1 << shifts) | (k1 >>> (28 - shifts));

         en_keysl[round] = comp_maskl[0][(t0 >>> 21) & 0x7f]
                         | comp_maskl[1][(t0 >>> 14) & 0x7f]
                         | comp_maskl[2][(t0 >>> 7) & 0x7f]
                         | comp_maskl[3][t0 & 0x7f]
                         | comp_maskl[4][(t1 >>> 21) & 0x7f]
                         | comp_maskl[5][(t1 >>> 14) & 0x7f]
                         | comp_maskl[6][(t1 >>> 7) & 0x7f]
                         | comp_maskl[7][t1 & 0x7f];

         en_keysr[round] = comp_maskr[0][(t0 >>> 21) & 0x7f]
                         | comp_maskr[1][(t0 >>> 14) & 0x7f]
                         | comp_maskr[2][(t0 >>> 7) & 0x7f]
                         | comp_maskr[3][t0 & 0x7f]
                         | comp_maskr[4][(t1 >>> 21) & 0x7f]
                         | comp_maskr[5][(t1 >>> 14) & 0x7f]
                         | comp_maskr[6][(t1 >>> 7) & 0x7f]
                         | comp_maskr[7][t1 & 0x7f];
 }
}

var saltbits;
function des_setup_salt(salt)
{
        saltbits = 0;
        saltbit = 1;
        obit = 0x800000;
        for (i = 0; i < 24; i++) {
                if (salt & saltbit)
                        saltbits |= obit;
                saltbit <<= 1;
                obit >>= 1;
        }
}

var des_r0, des_r1;

function des_do_des()
{
 var l, r, f, r48l, r48r;
 var count = 25;

 /* Don't bother with initial permutation. */
 l = r = 0;

 while (count--) {
         /* Do each round. */
         kl = 0;
         kr = 0;
         var round = 16;
         while (round--) {
                 /* Expand R to 48 bits (simulate the E-box). */
                 r48l    = ((r & 0x00000001) << 23)
                         | ((r & 0xf8000000) >>> 9)
                         | ((r & 0x1f800000) >>> 11)
                         | ((r & 0x01f80000) >>> 13)
                         | ((r & 0x001f8000) >>> 15);

                 r48r    = ((r & 0x0001f800) << 7)
                         | ((r & 0x00001f80) << 5)
                         | ((r & 0x000001f8) << 3)
                         | ((r & 0x0000001f) << 1)
                         | ((r & 0x80000000) >>> 31);
                 /*
                  * Do salting for crypt() and friends, and
                  * XOR with the permuted key.
                  */
                 f = (r48l ^ r48r) & saltbits;
                 r48l ^= f ^ en_keysl[ kl++ ];
                 r48r ^= f ^ en_keysr[ kr++ ];
                 /*
                  * Do sbox lookups (which shrink it back to 32 bits)
                  * and do the pbox permutation at the same time.
                  */
                 f = psbox[0][m_sbox[0][r48l >> 12]]
                   | psbox[1][m_sbox[1][r48l & 0xfff]]
                   | psbox[2][m_sbox[2][r48r >> 12]]
                   | psbox[3][m_sbox[3][r48r & 0xfff]];
                 /*
                  * Now that we've permuted things, complete f().
                  */
                 f ^= l;
                 l = r;
                 r = f;
         }
         r = l;
         l = f;
 }

 /* Final permutation (inverse of IP). */
 des_r0 = fp_maskl[0][l >>> 24]
        | fp_maskl[1][(l >>> 16) & 0xff]
        | fp_maskl[2][(l >>> 8) & 0xff]
        | fp_maskl[3][l & 0xff]
        | fp_maskl[4][r >>> 24]
        | fp_maskl[5][(r >>> 16) & 0xff]
        | fp_maskl[6][(r >>> 8) & 0xff]
        | fp_maskl[7][r & 0xff];
 des_r1 = fp_maskr[0][l >>> 24]
        | fp_maskr[1][(l >>> 16) & 0xff]
        | fp_maskr[2][(l >>> 8) & 0xff]
        | fp_maskr[3][l & 0xff]
        | fp_maskr[4][r >>> 24]
        | fp_maskr[5][(r >>> 16) & 0xff]
        | fp_maskr[6][(r >>> 8) & 0xff]
        | fp_maskr[7][r & 0xff];
}

function descrypt(key, salt_str)
{
        var keybuf = new Array(8);
        var output = salt_str.slice(0,2);

        q = 0;
        keypos = 0;
        while (q < 8) {
          keybuf[q] = key.charCodeAt(keypos) << 1;
          q++;
          if (keypos < key.length) keypos++;
        }

        des_setkey(keybuf);

        /* This is the "old style" DES crypt. */
        var salt = (ascii_to_bin(salt_str.charCodeAt(1)) << 6) |
                    ascii_to_bin(salt_str.charCodeAt(0));
        des_setup_salt(salt);
        des_do_des();

        l = (des_r0 >>> 8);
        output += ascii64.charAt( (l >>> 18) & 0x3f );
        output += ascii64.charAt( (l >>> 12) & 0x3f );
        output += ascii64.charAt( (l >>>  6) & 0x3f );
        output += ascii64.charAt( (l >>>  0) & 0x3f );

        l = (des_r0 << 16) | ((des_r1 >>> 16) & 0xffff);
        output += ascii64.charAt( (l >>> 18) & 0x3f );
        output += ascii64.charAt( (l >>> 12) & 0x3f );
        output += ascii64.charAt( (l >>>  6) & 0x3f );
        output += ascii64.charAt( (l >>>  0) & 0x3f );

        l = (des_r1 << 2);
        output += ascii64.charAt( (l >>> 12) & 0x3f );
        output += ascii64.charAt( (l >>>  6) & 0x3f );
        output += ascii64.charAt( (l >>>  0) & 0x3f );

        return output;
}

des_init();
